bundlewrap/metagen.py

```diff
@@ -19,6 +19,180 @@
 MAX_METADATA_ITERATIONS = int(environ.get("BW_MAX_METADATA_ITERATIONS", "1000"))
 
 
+class PathSet:
+    """
+    Collects metadata paths and stores only the highest levels ones.
+
+    >>> s = PathSet()
+    >>> s.add(("foo", "bar"))
+    >>> s.add(("foo",))
+    >>> s
+    {"foo"}
+    """
+
+    def __init__(self):
+        self._paths = set()
+
+    def __iter__(self):
+        for path in self._paths:
+            yield path
+
+    def __len__(self):
+        return len(self._paths)
+
+    def __repr__(self):
+        return "<PathSet: {}>".format(repr(self._paths))
+
+    def add(self, new_path):
+        if self.covers(new_path):
+            return False
+        for existing_path in self._paths.copy():
+            if list_starts_with(existing_path, new_path):
+                self._paths.remove(existing_path)
+        self._paths.add(new_path)
+        return True
+
+    def covers(self, candidate_path):
+        for existing_path in self._paths:
+            if list_starts_with(candidate_path, existing_path):
+                return True
+        return False
+
+
+def reactors_for_paths(available_reactors, required_paths):
+    """
+    Returns only those available_reactors that might affect the
+    required_paths.
+    """
+    for name, reactor in available_reactors:
+        provides = getattr(reactor, '_provides', tuple())
+        if provides:
+            for path in provides:
+                if required_paths.covers(path):
```

Comment:
Shouldn’t this be the other way around? The path from the decorator should cover one of the required paths?

Reply:
It's actually both ways:

```
metadata.get('foo') -> required_paths == {'foo'}
@metadata_reactor.provides('foo/bar') -> foo covers foo/bar, so this reactor is relevant
```

```
metadata.get('foo/bar') -> required_paths == {'foo/bar'}
@metadata_reactor.provides('foo') -> reactor might provide foo/bar, so it's relevant
```

But you're right, the latter case is not handled properly.

Reply:
pushed a quick fix

Reply:
Aha! Crashes gone. Metadata hash remained the same. 🎉